File: easy_multi_provider/usage_history.py
"""Incremental, read-only Codex rollout accounting. Never retain message bodies."""

from __future__ import annotations

import hashlib
import json
import threading
import time
from datetime import datetime
from pathlib import Path

from .performance import token_count
from .usage_ledger import TOKEN_FIELDS, usage_identity

_FIELDS = dict(zip(TOKEN_FIELDS, ("input_tokens", "output_tokens", "cached_input_tokens",
                               "cache_write_input_tokens", "cache_write_1h_tokens", "reasoning_output_tokens")))


def _timestamp(value):
    try:
        parsed = datetime.fromisoformat(str(value).replace("Z", "+00:00"))
        return parsed.timestamp() if parsed.tzinfo is not None else None
    except (ValueError, OverflowError, OSError):
        return None


def _counts(value):
    if not isinstance(value, dict):
        return None
    result = {field: value.get(field) for field in _FIELDS.values()}
    # Cumulative counters can exceed the per-request size limit.
    return {key: number for key, number in result.items()
            if type(number) is int and 0 <= number < 2**63}


def history_routes(config):
    """Current routes suggest a category, not the identity of an old login."""
    providers = {item["id"]: item for item in config.get("providers", [])}
    routes = {}
    for account in config.get("accounts", []):
        if account.get("prefix"):
            routes[str(account["prefix"]) + "/"] = ("subscription", "")
    for model in config.get("models", []):
        provider = providers.get(model.get("provider"))
        if provider:
            identity = usage_identity(provider, model)
            routes[model["id"]] = (identity["usage_category"], identity["upstream_model"])
    return routes


def parse_record(record, state, routes):
    """Advance a small numeric cursor and return at most one usage observation."""
    payload = record.get("payload")
    if not isinstance(payload, dict):
        return None
    kind = record.get("type")
    if kind == "session_meta":
        state.update(thread=str(payload.get("id") or payload.get("session_id") or ""),
                     provider=str(payload.get("model_provider") or ""),
                     fork=str(payload.get("forked_from_id") or ""))
        return None
    if kind == "turn_context":
        state["model"] = str(payload.get("model") or "")[:256]
        state["turn"] = str(payload.get("turn_id") or state.get("turn") or "")
        state["tier"] = str(payload.get("service_tier") or "default")
        return None
    if kind != "event_msg":
        return None
    if payload.get("type") == "task_started":
        state["turn"] = str(payload.get("turn_id") or "")
        return None
    if payload.get("type") != "token_count" or not isinstance(payload.get("info"), dict):
        return None
    info = payload["info"]
    totals, last = _counts(info.get("total_token_usage")), _counts(info.get("last_token_usage"))
    previous = state.get("totals")
    if totals and totals == previous:
        return None  # rate-limit-only notifications repeat the last usage
    state["totals"] = totals
    usage = last
    issue = None
    if not usage and totals:
        if previous and all(totals.get(key, 0) >= previous.get(key, 0) for key in totals):
            usage = {key: count - previous.get(key, 0) for key, count in totals.items()}
        elif not state.get("fork"):
            usage = totals
        # A cumulative-only observation may contain several model calls; long
        # context pricing cannot safely be applied to it as a single request.
        issue = "aggregate_usage"
    if not usage:
        return None
    model = state.get("model", "")
    observed_at = _timestamp(record.get("timestamp"))
    if not model or observed_at is None:
        state["unattributed"] = state.get("unattributed", 0) + 1
        return None
    category, upstream = "unknown", model
    if "/" not in model and state.get("provider") == "openai" and model.startswith(("gpt-", "codex-")):
        category = "native"
    elif model in routes:
        category, upstream = routes[model]
    elif "/" in model and model.split("/", 1)[0] + "/" in routes:
        category = routes[model.split("/", 1)[0] + "/"][0]
        upstream = model.split("/", 1)[1]
    owner = "history:" + (model.split("/", 1)[0] if "/" in model else state.get("provider", "unknown"))
    event = {field: token_count(usage[source]) for field, source in _FIELDS.items() if source in usage}
    raw_last = info.get("last_token_usage") or {}
    total = raw_last.get("total_tokens")
    if (type(total) is int and event.get("input_tokens") is not None and event.get("output_tokens") is not None
            and total != event["input_tokens"] + event["output_tokens"]):
        issue = "inconsistent_usage"
    if (event.get("reasoning_tokens") or 0) > (event.get("output_tokens") or 0):
        issue = "inconsistent_usage"
    identity = [state.get("turn") or state.get("thread"), record.get("timestamp"), model, usage]
    row_id = "history:" + hashlib.sha256(json.dumps(identity, sort_keys=True).encode()).hexdigest()
    event.update(observation_id=row_id, origin="history", usage_turn=state.get("turn", ""),
                 route_model=model, route="responses", usage_category=category, usage_owner=owner,
                 upstream_model=upstream, service_tier=state.get("tier", "default"), usage_issue=issue)
    return event, observed_at


class UsageHistoryScanner:
    """One worker; file checkpoints commit with their observations in SQLite."""

    def __init__(self, ledger, codex_home, config):
        self.ledger, self.codex_home, self.config = ledger, codex_home, config
        self._stop = threading.Event()
        self._wake = threading.Event()
        self._scan_lock = threading.Lock()
        self._thread = None
        self.status = {"running": False, "files": 0, "updated": 0, "errors": 0, "last_scan_at": None}

    def scan(self):
        if not self._scan_lock.acquire(blocking=False):
            return
        self.status = {**self.status, "running": True, "queued": False, "files": 0, "updated": 0, "errors": 0}
        try:
            home = Path(self.codex_home()).resolve()
            routes = history_routes(self.config())
            for directory in (home / "sessions", home / "archived_sessions"):
                if not directory.is_dir() or directory.is_symlink():
                    continue
                for path in directory.rglob("*.jsonl"):
                    if self._stop.is_set():
                        return
                    self.status["files"] += 1
                    try:
                        if path.is_symlink():
                            continue
                        path.resolve().relative_to(home)
                        if self._scan_file(path, routes):
                            self.status["updated"] += 1
                    except (OSError, ValueError, TypeError):
                        self.status["errors"] += 1
            self.status["last_scan_at"] = time.time()
        except Exception as exc:
            self.status["errors"] += 1
            self.ledger.journal.event("warning", "usage_history_scan_failed", exception_type=type(exc).__name__)
        finally:
            self.status["running"] = False
            self._scan_lock.release()

    def _scan_file(self, path, routes):
        stat = path.stat()
        saved = self.ledger.history_checkpoint(str(path))
        # ctime changes on append on Unix. The inode plus a prefix digest
        # distinguishes replacements while allowing append-only progress.
        with path.open("rb") as stream:
            prefix = hashlib.sha256(stream.read(min(256, stat.st_size))).hexdigest()
            identity = [stat.st_dev, stat.st_ino, prefix]
            def tail_digest(offset):
                position = stream.tell()
                stream.seek(max(0, offset - 256))
                digest = hashlib.sha256(stream.read(min(256, offset))).hexdigest()
                stream.seek(position)
                return digest

            if (saved and saved["identity"] == identity and stat.st_size >= saved["offset"]
                    and saved.get("tail") == tail_digest(saved["offset"])):
                if stat.st_size == saved["size"] and stat.st_mtime_ns == saved["mtime"]:
                    return False
                state, offset = saved["state"], saved["offset"]
            else:
                state, offset = {}, 0
            stream.seek(offset)
            events = []
            while not self._stop.is_set():
                line = stream.readline()
                if not line or not line.endswith(b"\n"):
                    break  # leave an incomplete final JSON object for the next scan
                offset = stream.tell()
                if not any(marker in line for marker in (b'"session_meta"', b'"turn_context"',
                                                         b'"token_count"', b'"task_started"')):
                    continue
                try:
                    record = json.loads(line)
                    parsed = parse_record(record, state, routes) if isinstance(record, dict) else None
                    if parsed:
                        events.append(parsed)
                except (ValueError, TypeError, OverflowError):
                    state["malformed"] = state.get("malformed", 0) + 1
                if len(events) >= 200:
                    self.ledger.record_history(events, str(path), {"identity": identity, "offset": offset,
                        "size": offset, "mtime": stat.st_mtime_ns, "state": state, "tail": tail_digest(offset)})
                    events.clear()
            self.ledger.record_history(events, str(path), {"identity": identity, "offset": offset,
                "size": stat.st_size, "mtime": stat.st_mtime_ns, "state": state, "tail": tail_digest(offset)})
        return True

    def _run(self):
        while not self._stop.is_set():
            self._wake.clear()
            self.scan()
            self._wake.wait(60)

    def start(self):
        if self._thread is None:
            self._thread = threading.Thread(target=self._run, daemon=True, name="emp-usage-history")
            self._thread.start()

    def refresh(self):
        self.status["queued"] = True
        self.start()
        self._wake.set()

    def stop(self):
        self._stop.set()
        self._wake.set()
        if self._thread is not None:
            self._thread.join(timeout=2)

File: crates/emp-codex/src/usage_history.rs
//! Incremental read-only usage scanning of Codex rollout files.
use emp_state::usage::{
    TOKEN_FIELDS,
    ledger::{UsageError, UsageLedger},
    token_count,
};
use serde_json::{Map, Value, json};
use sha2::{Digest, Sha256};
use std::collections::BTreeMap;
use std::fs::File;
use std::io::{BufRead, BufReader, Read, Seek, SeekFrom};
use std::path::Path;
use std::sync::{
    Mutex,
    atomic::{AtomicBool, Ordering},
};
use time::{OffsetDateTime, format_description::well_known::Rfc3339};
const COUNTS: [&str; 6] = [
    "input_tokens",
    "output_tokens",
    "cached_input_tokens",
    "cache_write_input_tokens",
    "cache_write_1h_tokens",
    "reasoning_output_tokens",
];
type Routes = BTreeMap<String, (String, String)>;
fn text(value: &Value, key: &str) -> String {
    value[key].as_str().unwrap_or("").to_owned()
}
fn counts(value: &Value) -> Option<Value> {
    value.as_object().map(|value| {
        Value::Object(
            COUNTS
                .into_iter()
                .filter_map(|key| {
                    value
                        .get(key)
                        .and_then(Value::as_i64)
                        .filter(|n| *n >= 0)
                        .map(|n| (key.into(), json!(n)))
                })
                .collect(),
        )
    })
}
fn routes(config: &Value) -> Routes {
    let mut result = Routes::new();
    for account in config["accounts"].as_array().into_iter().flatten() {
        if let Some(prefix) = account["prefix"].as_str().filter(|s| !s.is_empty()) {
            result.insert(format!("{prefix}/"), ("subscription".into(), String::new()));
        }
    }
    for model in config["models"].as_array().into_iter().flatten() {
        if let Some(provider) = config["providers"]
            .as_array()
            .into_iter()
            .flatten()
            .find(|p| p["id"] == model["provider"])
        {
            let category = match provider["auth_mode"].as_str() {
                Some("account") => "subscription",
                Some("forward" | "native") => "native",
                _ => "external",
            };
            let upstream = model["upstream_id"]
                .as_str()
                .filter(|s| !s.is_empty())
                .unwrap_or_else(|| model["id"].as_str().unwrap_or(""));
            let provider_prefix = format!("{}/", provider["id"].as_str().unwrap_or(""));
            let upstream = upstream.strip_prefix(&provider_prefix).unwrap_or(upstream);
            result.insert(text(model, "id"), (category.into(), upstream.into()));
        }
    }
    result
}
fn parse_record(record: &Value, state: &mut Value, routes: &Routes) -> Option<(Value, f64)> {
    let payload = record.get("payload").filter(|value| value.is_object())?;
    match record["type"].as_str()? {
        "session_meta" => {
            state["thread"] = json!(
                payload
                    .get("id")
                    .or_else(|| payload.get("session_id"))
                    .and_then(Value::as_str)
                    .unwrap_or("")
            );
            state["provider"] = json!(text(payload, "model_provider"));
            state["fork"] = json!(text(payload, "forked_from_id"));
            return None;
        }
        "turn_context" => {
            state["model"] = json!(text(payload, "model").chars().take(256).collect::<String>());
            if payload["turn_id"].as_str().is_some_and(|s| !s.is_empty()) {
                state["turn"] = payload["turn_id"].clone();
            }
            state["tier"] = json!(
                payload["service_tier"]
                    .as_str()
                    .filter(|s| !s.is_empty())
                    .unwrap_or("default")
            );
            return None;
        }
        "event_msg" => {}
        _ => return None,
    }
    if payload["type"] == "task_started" {
        state["turn"] = json!(text(payload, "turn_id"));
        return None;
    }
    if payload["type"] != "token_count" {
        return None;
    }
    let info = payload.get("info").filter(|value| value.is_object())?;
    let totals = counts(&info["total_token_usage"]).unwrap_or(Value::Null);
    let previous = state["totals"].clone();
    if totals.as_object().is_some_and(|v| !v.is_empty()) && totals == previous {
        return None;
    }
    state["totals"] = totals.clone();
    let mut usage = counts(&info["last_token_usage"]).unwrap_or(Value::Null);
    let mut issue = Value::Null;
    if usage.as_object().is_none_or(|v| v.is_empty())
        && let Some(totals) = totals.as_object().filter(|v| !v.is_empty())
    {
        if previous.as_object().is_some_and(|v| !v.is_empty())
            && totals.iter().all(|(key, value)| {
                value.as_i64().unwrap_or(0) >= previous[key].as_i64().unwrap_or(0)
            })
        {
            usage = Value::Object(
                totals
                    .iter()
                    .map(|(key, value)| {
                        (
                            key.clone(),
                            json!(
                                value.as_i64().unwrap_or(0) - previous[key].as_i64().unwrap_or(0)
                            ),
                        )
                    })
                    .collect(),
            );
        } else if state["fork"].as_str().is_none_or(str::is_empty) {
            usage = Value::Object(totals.clone());
        }
        issue = json!("aggregate_usage");
    }
    if usage.as_object().is_none_or(|v| v.is_empty()) {
        return None;
    }
    let model = text(state, "model");
    let timestamp = record["timestamp"]
        .as_str()
        .and_then(|value| OffsetDateTime::parse(value, &Rfc3339).ok());
    if model.is_empty() || timestamp.is_none() {
        state["unattributed"] = json!(state["unattributed"].as_u64().unwrap_or(0) + 1);
        return None;
    }
    let mut category = "unknown".to_owned();
    let mut upstream = model.clone();
    if !model.contains('/')
        && state["provider"] == "openai"
        && (model.starts_with("gpt-") || model.starts_with("codex-"))
    {
        category = "native".into();
    } else if let Some((kind, name)) = routes.get(&model) {
        category = kind.clone();
        upstream = name.clone();
    } else if let Some((prefix, tail)) = model.split_once('/')
        && let Some((kind, _)) = routes.get(&format!("{prefix}/"))
    {
        category = kind.clone();
        upstream = tail.into();
    }
    let owner = format!(
        "history:{}",
        model
            .split_once('/')
            .map(|(prefix, _)| prefix)
            .unwrap_or_else(|| state["provider"].as_str().unwrap_or("unknown"))
    );
    let mut event = Map::new();
    for (field, source) in TOKEN_FIELDS.into_iter().zip(COUNTS) {
        if let Some(value) = usage.get(source) {
            event.insert(field.into(), json!(token_count(Some(value))));
        }
    }
    if let (Some(total), Some(input), Some(output)) = (
        info["last_token_usage"]["total_tokens"].as_i64(),
        event.get("input_tokens").and_then(Value::as_i64),
        event.get("output_tokens").and_then(Value::as_i64),
    ) && total != input + output
    {
        issue = json!("inconsistent_usage");
    }
    if event
        .get("reasoning_tokens")
        .and_then(Value::as_u64)
        .unwrap_or(0)
        > event
            .get("output_tokens")
            .and_then(Value::as_u64)
            .unwrap_or(0)
    {
        issue = json!("inconsistent_usage");
    }
    let turn = state
        .get("turn")
        .filter(|value| value.as_str().is_some_and(|s| !s.is_empty()))
        .or_else(|| state.get("thread"))
        .cloned()
        .unwrap_or(Value::Null);
    let row_id = format!(
        "history:{}",
        emp_state::usage::identity_hash(&json!([turn, record["timestamp"], model, usage]))
    );
    event.extend(json!({"observation_id":row_id,"origin":"history","usage_turn":text(state,"turn"),"route_model":model,"route":"responses","usage_category":category,"usage_owner":owner,"upstream_model":upstream,"service_tier":state["tier"].as_str().unwrap_or("default"),"usage_issue":issue}).as_object().unwrap().clone());
    Some((
        Value::Object(event),
        timestamp.unwrap().unix_timestamp_nanos() as f64 / 1e9,
    ))
}

pub struct UsageHistoryScanner {
    status: Mutex<Value>,
    scan_lock: Mutex<()>,
    pub stop: AtomicBool,
}
impl Default for UsageHistoryScanner {
    fn default() -> Self {
        Self {
            status: Mutex::new(
                json!({"running":false,"files":0,"updated":0,"errors":0,"last_scan_at":null}),
            ),
            scan_lock: Mutex::new(()),
            stop: AtomicBool::new(false),
        }
    }
}
impl UsageHistoryScanner {
    pub fn status(&self) -> Value {
        self.status.lock().expect("usage history status").clone()
    }
    pub fn queued(&self) {
        self.status.lock().expect("usage history status")["queued"] = json!(true);
    }
    fn increment(&self, key: &str) {
        let mut state = self.status.lock().expect("usage history status");
        state[key] = json!(state[key].as_u64().unwrap_or(0) + 1);
    }
    pub fn scan(&self, ledger: &UsageLedger, home: &Path, config: &Value, now: f64) {
        let Ok(_lock) = self.scan_lock.try_lock() else {
            return;
        };
        {
            let mut status = self.status.lock().expect("usage history status");
            for (key, value) in [
                ("running", json!(true)),
                ("queued", json!(false)),
                ("files", json!(0)),
                ("updated", json!(0)),
                ("errors", json!(0)),
            ] {
                status[key] = value;
            }
        }
        let routes = routes(config);
        let home = emp_state::config::resolve_user_path(home);
        // Compare paths in the same representation. Windows canonicalize()
        // adds the extended-length prefix to rollout paths, which a resolved
        // CODEX_HOME path does not necessarily have.
        let home = home.canonicalize().unwrap_or(home);
        for directory in [home.join("sessions"), home.join("archived_sessions")] {
            self.scan_directory(ledger, &home, &directory, &routes);
        }
        let mut status = self.status.lock().expect("usage history status");
        status["last_scan_at"] = json!(now);
        status["running"] = json!(false);
    }
    fn scan_directory(&self, ledger: &UsageLedger, home: &Path, directory: &Path, routes: &Routes) {
        if directory.is_symlink() || self.stop.load(Ordering::Acquire) {
            return;
        }
        let Ok(entries) = directory.read_dir() else {
            return;
        };
        for entry in entries.flatten() {
            if self.stop.load(Ordering::Acquire) {
                return;
            }
            let path = entry.path();
            if path.is_dir() {
                self.scan_directory(ledger, home, &path, routes);
                continue;
            }
            if path.extension().is_none_or(|ext| ext != "jsonl") {
                continue;
            }
            self.increment("files");
            if path.is_symlink() || !path.is_file() {
                continue;
            }
            if path
                .canonicalize()
                .is_ok_and(|resolved| resolved.starts_with(home))
            {
                match self.scan_file(ledger, &path, routes) {
                    Ok(true) => self.increment("updated"),
                    Ok(false) => {}
                    Err(_) => self.increment("errors"),
                }
            } else {
                self.increment("errors");
            }
        }
    }
    fn scan_file(
        &self,
        ledger: &UsageLedger,
        path: &Path,
        routes: &Routes,
    ) -> Result<bool, UsageError> {
        let metadata = path.metadata()?;
        let size = metadata.len();
        let mtime = metadata
            .modified()?
            .duration_since(std::time::UNIX_EPOCH)
            .map_err(|_| UsageError)?
            .as_nanos();
        let saved = ledger.history_checkpoint(&path.to_string_lossy())?;
        let mut stream = BufReader::new(File::open(path)?);
        let prefix = digest(&mut stream, 0, size.min(256))?;
        #[cfg(unix)]
        let identity = {
            use std::os::unix::fs::MetadataExt;
            json!([metadata.dev(), metadata.ino(), prefix])
        };
        #[cfg(not(unix))]
        let identity = json!([
            0,
            metadata
                .created()
                .ok()
                .and_then(|time| time.duration_since(std::time::UNIX_EPOCH).ok())
                .map(|time| time.as_nanos().to_string()),
            prefix
        ]);
        let mut offset = 0;
        let mut state = json!({});
        if let Some(saved) = &saved
            && saved["identity"] == identity
            && let Some(position) = saved["offset"].as_u64().filter(|offset| *offset <= size)
            && saved["tail"]
                == digest(&mut stream, position.saturating_sub(256), position.min(256))?
        {
            if saved["size"] == size && saved["mtime"].as_u64().map(u128::from) == Some(mtime) {
                return Ok(false);
            }
            state = saved["state"].clone();
            offset = position;
        }
        stream.seek(SeekFrom::Start(offset))?;
        let mut events = Vec::new();
        let mut line = Vec::new();
        while !self.stop.load(Ordering::Acquire) {
            line.clear();
            if stream.read_until(b'\n', &mut line)? == 0 || !line.ends_with(b"\n") {
                break;
            }
            offset = stream.stream_position()?;
            if ![
                b"\"session_meta\"".as_slice(),
                b"\"turn_context\"",
                b"\"token_count\"",
                b"\"task_started\"",
            ]
            .iter()
            .any(|marker| line.windows(marker.len()).any(|window| window == *marker))
            {
                continue;
            }
            match serde_json::from_slice::<Value>(&line) {
                Ok(record) => {
                    if let Some(event) = parse_record(&record, &mut state, routes) {
                        events.push(event);
                    }
                }
                Err(_) => state["malformed"] = json!(state["malformed"].as_u64().unwrap_or(0) + 1),
            }
            if events.len() >= 200 {
                let checkpoint = json!({"identity":identity,"offset":offset,"size":offset,"mtime":mtime as u64,"state":state,"tail":digest(&mut stream,offset.saturating_sub(256),offset.min(256))?});
                ledger.record_history(&events, &path.to_string_lossy(), &checkpoint)?;
                events.clear();
            }
        }
        let checkpoint = json!({"identity":identity,"offset":offset,"size":size,"mtime":mtime as u64,"state":state,"tail":digest(&mut stream,offset.saturating_sub(256),offset.min(256))?});
        ledger.record_history(&events, &path.to_string_lossy(), &checkpoint)?;
        Ok(true)
    }
}
fn digest(stream: &mut BufReader<File>, start: u64, length: u64) -> std::io::Result<String> {
    let position = stream.stream_position()?;
    stream.seek(SeekFrom::Start(start))?;
    let mut bytes = vec![0; length as usize];
    stream.read_exact(&mut bytes)?;
    stream.seek(SeekFrom::Start(position))?;
    Ok(format!("{:x}", Sha256::digest(bytes)))
}
